### DISCUSSION TITLE
How can I match ~/.bash* files but exclude ~/.bash_history?

### DISCUSSION THREAD

**Discussion by OnlineCop** [BurntSushi/ripgrep#2557] [Q&A] [answered]
# Question (TL;DR version)
Is there a way for a RipGrep glob filter (`-g, --glob`) to override explicitly-included command-name filenames (which may be the result of Bash globbing)?

# Simple Example

When I couple a `-g` glob filter with filenames that the glob filter *would have excluded*, those files are still included in the search.

```bash
$ rg -c -g'!exclude_me.txt' 'text' ./exclude_me.txt ./include_me.txt
./exclude_me.txt:2
./include_me.txt:3
```

While the above example explicitly lists `exclude_me.txt` in the files to search, this is what RipGrep sees due to Bash glob expansion with a command such as:

```bash
$ rg -c -g'!exclude_me.txt' 'text' ./*.txt
```

# Deeper Dive

My $HOME directory contains several files (and one directory) starting with "`.bash*`":
1. `.bash_sessions/` (directory containing 10+ files)
1. `.bash_history` (file I wish to exclude)
1. `.bashrc`
1. `.bash_profile`
1. `.bash_aliases`
1. `.bash_regexes` (does not contain "text")
1. `.bash_functions`

I would like to search through all these *except* # 1 (the `.bash_sessions` subdirectory containing other files) or # 2 (`.bash_history`).

1. Without any glob filtering:
    ```bash
    $ rg -c 'text' ~/.bash*
    /Users/user/.bash_sessions/file1:1
    /Users/user/.bash_sessions/file2:6
    /Users/user/.bash_sessions/file6:3
    /Users/user/.bash_history:90
    /Users/user/.bashrc:2
    /Users/user/.bash_profile:3
    /Users/user/.bash_aliases:1
    /Users/user/.bash_functions:1
    ```

    I wish to exclude those first 4 results (`.bash_sessions/*` and `.bash_history`).

1. Using a glob filter to remove the subdirectory:
    ```bash
    $ rg -c -g'!**/.bash_sessions/*' 'text' ~/.bash*
    /Users/user/.bash_history:90
    /Users/user/.bashrc:2
    /Users/user/.bash_profile:3
    /Users/user/.bash_aliases:1
    /Users/user/.bash_functions:1
    ```

    Good so far, but I still want to exclude the first result (`.bash_history`).

1. I have tried appending the following globs to exclude `.bash_history` but none seem to have an effect:
    ```bash
    $ rg -c -g'!**/.bash_sessions/*' [globs found below] 'text' ~/.bash*
    ```

    1. `-g'!.bash_history'`
    1. `-g'!*.bash_history'`
    1. `-g'!*/.bash_history'`
    1. `-g'!**.bash_history'`
    1. `-g'!**/.bash_history'`
    1. `-g'!**/*.bash_history'`
    1. `-g'!**/*/.bash_history'`
    1. `-g'!~.bash_history'`        or `-g'!'"$HOME"'.bash_history`
    1. `-g'!~/.bash_history'`       or `-g'!'"$HOME"'/.bash_history'`
    1. `-g'!*/~/.bash_history'`     or `-g'!*/'"$HOME"'/.bash_history'`
    1. `-g'!**/~/.bash_history'`    or `-g'!**/'"$HOME"'/.bash_history'`

    The `.bash_history` file is included when any of the above were used. Lines 8-11 tried exchanging `~` for the full `$HOME` (aka `/Users/user`) path.

1. The following solution *partially* works, but has its own drawbacks:
    ```bash
    $ rg -c -g'.bash*' -g'!**/.bash_sessions/*' -g'!.bash_history' 'text' ~/
    /Users/user/SomeDirectory/.bash_completions:10
    /Users/user/AnotherDirectory/.bashrc~
    /Users/user/.bashrc:2
    /Users/user/Library/SomeOffLimitDirectory: Operation not permitted (os error 1) #stderr
    /Users/user/.bash_profile:3
    /Users/user/.bash_aliases:1
    /Users/user/Library/AnotherOffLimitDirectory: Operation not permitted (os error 1) #stderr
    /Users/user/.bash_functions:1
    ```

    This prevents Bash globbing from explicitly passing `.bash_history` or `.bash_sessions` to RipGrep.

    Unfortunately, it recurses into OTHER subdirectories that are not children of `.bash_sessions`.

1. The above could be mitigated with either:
    1. `rg -c -g'.bash*' -g'!**/*/' -g'!.bash_history' 'text' ~/`
    1. `rg -c -g'.bash*' --maxdepth=1 -g'!.bash_history' 'text' ~/`

    But this prevents me from including files found in subdirectories from *other* paths.

# What Is the Right Way To Do This?

Should I simply be passing in different options, or am I using the `-g` glob filter incorrectly?

**Comment by OnlineCop** [comment]
As an addendum to the above: the suggestion to filter these results via `| rg -v '.bash_history'` or `| grep -v '.bash_history'` isn't what I'm after, since RipGrep *does* allow a lot of other features (such as `--replace` or `--pcre2` and `--multiline` searches), which does not always get properly filtered out with those methods.

**Comment by BurntSushi** [comment, marked as answer]
> Is there a way for a RipGrep glob filter (-g, --glob) to override explicitly-included command-name filenames (which may be the result of Bash globbing)?

No. This is by design. The driving design point here is that if you tell ripgrep to search a _specific_ file, then it will absolutely do so regardless of any other flags. ripgrep doesn't know about shell globbing as that's expanded before ripgrep itself is invoked. So there is no way for ripgrep to tell the difference between "explicitly specified file" and "file that matched a shell glob."

If I were in your situation, I'd probably just specify the files you want to search explicitly since the list is small:

```
$ rg text .bashrc .bash_profile .bash_aliases .bash_regexes .bash_functions
```

And then put that in an alias or a wrapper script if you do it frequently. Otherwise, you can always do things a bit more explicitly with `find` like this:

```
$ find ./.bash* | rg -v '.bash_(sessions|history)' | xargs rg text
```

Or to be pedantically correct:

```
$ find ./.bash* -print0 | rg -0 --null-data -v '.bash_(sessions|history)' | xargs -0 rg text
```
